new-packages/ts-project/src/state.ts

```diff
@@ -0,0 +1,203 @@
+  StateConfigDataWhitelistedProperties.has(key);
+
+export function extractState(
+  ctx: ExtractionContext,
+  ts: typeof import('typescript'),
+  state: Expression | undefined,
+  path: string[],
+): ExtractorStateConfig | undefined {
+  const result = {
+    states: [] as ExtractorStateConfig[],
+    data: {} as Partial<ExtractorStateConfig['data']>,
+  };
+
+  if (!state) {
+    return;
+  }
+
+  if (!ts.isObjectLiteralExpression(state)) {
+    ctx.errors.push({
+      type: 'state_unhandled',
+      node: state,
+    });
+    return;
+  }
+
+  for (const prop of state.properties) {
+    if (ts.isPropertyAssignment(prop)) {
+      const key = getPropertyKey(ctx, ts, prop);
+
+      /* state.states */
+      if (key === 'states' && ts.isObjectLiteralExpression(prop.initializer)) {
+        for (const state of prop.initializer.properties) {
+          if (ts.isPropertyAssignment(state)) {
+            const key = getPropertyKey(ctx, ts, state);
+            if (key) {
+              const stateConfig = extractState(
+                ctx,
+                ts,
+                state.initializer,
+                path.concat(key),
+              );
+              if (stateConfig) {
+                result.states.push(stateConfig);
+              }
+            }
+            continue;
+          }
+          if (ts.isShorthandPropertyAssignment(state)) {
+            ctx.errors.push({
+              type: 'state_property_unhandled',
+              node: state,
+            });
+            continue;
+          }
+          if (ts.isSpreadAssignment(state)) {
+            ctx.errors.push({
+              type: 'state_property_unhandled',
+              node: state,
+            });
+            continue;
+          }
+          if (
+            ts.isMethodDeclaration(state) ||
+            ts.isGetAccessorDeclaration(state) ||
+            ts.isSetAccessorDeclaration(state)
+          ) {
+            ctx.errors.push({
+              type: 'state_property_unhandled',
+              node: state,
+            });
+            continue;
+          }
+
+          state satisfies never;
+        }
+      } else if (key && isStateNodeConfigProperty(key)) {
+        switch (key) {
+          case 'initial': {
+            if (ts.isStringLiteral(prop.initializer)) {
+              result.data[key] = prop.initializer.text;
+            } else if (hasUnefinedValue(ts, prop.initializer)) {
+              result.data[key] = undefined;
+            } else {
+              ctx.errors.push({
+                type: 'state_property_unhandled',
+                node: prop,
+              });
+            }
+            break;
+          }
+          case 'type': {
+            const isStringWithValidValue =
+              ts.isStringLiteral(prop.initializer) &&
+              ['history', 'atomic', 'compound', 'parallel', 'final'].includes(
+                prop.initializer.text,
+              );
+
+            if (isStringWithValidValue) {
```

Comment:
More nesting would be preferred here, imho. First recognize the AST node - then handle its valid values. If you don't want to have specific errors in this branch of the code then `continue` when you managed to handle a thing successfully and otherwise just let it go to the unconditional `ctx.errors.push` below if/elseif